### DISCUSSION TITLE
Native Julia engine not working: "ERROR: Internal julia server error"

### DISCUSSION THREAD

**Discussion by mcanouil** [quarto-dev/quarto-cli#9551] [Q&A] [answered]
### Description

Using the latest development version and a very simple document:

````qmd
---
format: html
engine: julia
---

```{julia}
using Plots
plot(sin, 0, 2π)
```
````

The above produces:

```txt
Running [1/1] at line 6:  using Plots
ERROR: Julia server returned error after receiving "run" command:
Failed to run notebook: /Users/mcanouil/Projects/quarto/quarto-playground/index.qmd
ERROR: EvaluationError: Encountered 1 error during evaluation

Error 1 of 1
@ /Users/mcanouil/Projects/quarto/quarto-playground/index.qmd:6
ArgumentError: Package Plots not found in current path.
- Run `import Pkg; Pkg.add("Plots")` to install the Plots package.
Stacktrace:
  [1] macro expansion
    @ ./loading.jl:1772 [inlined]
  [2] macro expansion
    @ ./lock.jl:267 [inlined]
  [3] __require(into::Module, mod::Symbol)
    @ Base ./loading.jl:1753
  [4] #invoke_in_world#3
    @ ./essentials.jl:926 [inlined]
  [5] invoke_in_world
    @ ./essentials.jl:923 [inlined]
  [6] require(into::Module, mod::Symbol)
    @ Base ./loading.jl:1746
  [7] eval
    @ ./boot.jl:385 [inlined]
  [8] (::Main.var"#14#28"{Module, Expr})()
    @ Main ~/.julia/packages/QuartoNotebookRunner/pl36G/src/worker.jl:309
  [9] (::Main.IOCapture.var"#4#7"{DataType, Main.var"#14#28"{Module, Expr}, IOContext{Base.PipeEndpoint}, IOContext{Base.PipeEndpoint}, IOStream, Base.PipeEndpoint})()
    @ Main.IOCapture ~/.julia/packages/IOCapture/Rzdxd/src/IOCapture.jl:161
 [10] with_logstate(f::Function, logstate::Any)
    @ Base.CoreLogging ./logging.jl:515
 [11] with_logger
    @ ./logging.jl:627 [inlined]
 [12] capture(f::Main.var"#14#28"{Module, Expr}; rethrow::Type, color::Bool, passthrough::Bool, capture_buffer::IOBuffer)
    @ Main.IOCapture ~/.julia/packages/IOCapture/Rzdxd/src/IOCapture.jl:158
 [13] include_str(mod::Module, code::String; file::String, line::Int64)
    @ Main ~/.julia/packages/QuartoNotebookRunner/pl36G/src/worker.jl:293
 [14] #invokelatest#2
    @ ./essentials.jl:894 [inlined]
 [15] invokelatest
    @ ./essentials.jl:889 [inlined]
 [16] #7
    @ ~/.julia/packages/QuartoNotebookRunner/pl36G/src/worker.jl:164 [inlined]
 [17] with_inline_display(f::Main.var"#7#24"{String, String, Int64}, cell_options::Dict{String, Any})
    @ Main ~/.julia/packages/QuartoNotebookRunner/pl36G/src/worker.jl:53
 [18] _render_thunk(thunk::Function, code::String, cell_options::Dict{String, Any})
    @ Main ~/.julia/packages/QuartoNotebookRunner/pl36G/src/worker.jl:184
 [19] render(code::String, file::String, line::Int64, cell_options::Dict{String, Any})
    @ Main ~/.julia/packages/QuartoNotebookRunner/pl36G/src/worker.jl:161
 [20] top-level scope
    @ none:1

ERROR: Internal julia server error

Stack trace:
    at writeJuliaCommand (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/execute/julia.ts:663:13)
    at eventLoopTick (ext:core/01_core.js:153:7)
    at async executeJulia (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/execute/julia.ts:511:20)
    at async Object.execute (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/execute/julia.ts:137:16)
    at async renderExecute (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/command/render/render-files.ts:232:25)
    at async renderFileInternal (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/command/render/render-files.ts:592:35)
    at async renderFiles (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/command/render/render-files.ts:325:9)
    at async render (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/command/render/render-shared.ts:102:18)
    at async renderForPreview (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/command/preview/preview.ts:422:24)
    at async render (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/command/preview/preview.ts:166:22)
```

The error is very cryptic as `Plots` is installed.
<img width="1624" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/34bff5eb-9ea3-4a54-b8fc-d228289e2c18">

I am on macOS (latest).
Using Julia 1.10.2 (installed via Homebrew).

Quarto private environment:
```txt
~/Library/Caches/quarto/julia
├── Manifest.toml
├── Project.toml
└── julia_transport.txt
```

<img width="1624" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/b8ae161d-1551-4c3d-98b1-ef5c2e2b2c8d">

I don't see anything obvious here.

Even the following document does not work:

````qmd
---
format: html
engine: julia
---

```{julia}
versioninfo()
```
````

````txt
Running [1/1] at line 6:  versioninfo()
ERROR: Julia server returned error after receiving "run" command:
Failed to run notebook: /Users/mcanouil/Projects/quarto/quarto-playground/index.qmd
ERROR: EvaluationError: Encountered 1 error during evaluation

Error 1 of 1
@ /Users/mcanouil/Projects/quarto/quarto-playground/index.qmd:6
UndefVarError: `versioninfo` not defined
Stacktrace:
 [1] top-level scope
   @ ~/Projects/quarto/quarto-playground/index.qmd:7

ERROR: Internal julia server error

Stack trace:
    at writeJuliaCommand (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/execute/julia.ts:663:13)
    at eventLoopTick (ext:core/01_core.js:153:7)
    at async executeJulia (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/execute/julia.ts:511:20)
    at async Object.execute (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/execute/julia.ts:137:16)
    at async renderExecute (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/command/render/render-files.ts:232:25)
    at async renderFileInternal (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/command/render/render-files.ts:592:35)
    at async renderFiles (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/command/render/render-files.ts:325:9)
    at async render (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/command/render/render-shared.ts:102:18)
    at async renderForPreview (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/command/preview/preview.ts:422:24)
    at async render (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/command/preview/preview.ts:166:22)
````

While it works via Jupyter
<img width="1624" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/3c823254-3aa3-407a-b989-7b5cc266b18d">

Edit: I upgraded to 1.10.3 but I've got the exact same results.

**Comment by mcanouil** [comment]
@MichaelHatherly and @jkrumbiegel (cc @cscheid) if you have any pointers here.

**Comment by jkrumbiegel** [comment]
Two different things going on here:

I assume you have `Plots` installed in your main environment `@v1.10`. We have configured QuartoNotebookRunner to use the directory of the notebook as the default environment because this encourages using separate environments for separate projects. If you really want to use your main env from a notebook (which again is not advisable as you're bound to later make changes to it which break your notebook) you can set this in the frontmatter:

```yaml
julia:
    exeflags: ["--project=@v#.#"]
```

The `versioninfo` thing is happening because Julia pulls in some stdlib functions only when it's used in interactive mode (with a REPL). The same thing happens when you directly run some code with the `julia` binary:

```julia
% julia -E 'versioninfo()'
ERROR: UndefVarError: `versioninfo` not defined
Stacktrace:
 [1] top-level scope
   @ none:1
```

In this case it's the module `InteractiveUtils` so you'd have to call `using InteractiveUtils` first:

```julia
julia> @which versioninfo
InteractiveUtils
```

**Reply by mcanouil** [reply]
Thanks for the clarification/explanation.

We need to clearly document that when using the new native Julia engine, a Quarto project-level Julia library is required.
This is quite different from when using Jupyter.
Currently, there is only a small note on the option you just described, see <https://quarto.org/docs/computations/julia.html#using-the-julia-engine>.

Regarding the error message (not found in path), may I suggest to update it to clearly instruct the users that the engine is looking at a Quarto project-level library. Additionally, I would add instructions to fix the issue, by adding how to setup local library or that the users needs to set the flag/option to use the user-wide library.

**Reply by jkrumbiegel** [reply]
I can add a section to the docs that explains the project behavior more clearly. The error message itself we have no direct control over, unless we'd specifically intercept that one in QuartoNotebookRunner. Maybe something with `Experimental.register_error_hint` could work though...

**Reply by mcanouil** [reply]
The idea about the error message is because it's more direct for the users, especially if they come from Jupyter where it was working.

**Reply by mcanouil** [reply]
I am clearly missing something.
I've set (or at least I believe I did) a Julia environment:
<img width="950" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/132005ec-b9c0-4b09-ae6f-716e20d0d7b0">

But still, there seems to be something to be done somewhere as it still exhibits the same issue.

The installation section is not very clear and "Using custom versions of QuartoNotebookRunner" adds more confusion to the process.
It would be better to use code blocks to show the context and the command, whilst also allowing easier copy/pasting.

````qmd
```{.julia filename="Julia"}
]activate /some/dir
```
````

FYI, the above is very missleading for users not used to Julia, as you actually needs to "type" (not paste) `]` then paste or type the remaining of the command.

````qmd
```{.bash filename="Terminal"}
QUARTO_JULIA_PROJECT=/some/dir quarto render
```
````

**Reply by MichaelHatherly** [reply]
<img width="419" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/6144086/8dbbb2e1-814b-4aaa-b68a-e43fb0b62b0e">

What does `Pkg.status()` output for this REPL session?

**Reply by mcanouil** [reply]
```txt
julia> Pkg.status()
Status `~/Projects/quarto/quarto-playground/Project.toml`
  [91a5bcdd] Plots v1.40.4
```

**Reply by MichaelHatherly** [reply]
And if you run `Pkg.status()` in the quarto notebook that was failing with 

```
ArgumentError: Package Plots not found in current path.
- Run `import Pkg; Pkg.add("Plots")` to install the Plots package.
```

what is it's output? (Comment out the `using Plots` so it doesn't fail of course.)

**Reply by mcanouil** [reply]
```txt
❯ quarto preview /Users/mcanouil/Projects/quarto/quarto-playground/index.qmd --no-browser --no-watch-inputs
Running [1/1] at line 6:  Pkg.status()
ERROR: Julia server returned error after receiving "run" command:
Failed to run notebook: /Users/mcanouil/Projects/quarto/quarto-playground/index.qmd
ERROR: EvaluationError: Encountered 1 error during evaluation

Error 1 of 1
@ /Users/mcanouil/Projects/quarto/quarto-playground/index.qmd:6
no active project
Stacktrace:
 [1] pkgerror(msg::String)
   @ Pkg.Types /Applications/Julia-1.10.app/Contents/Resources/julia/share/julia/stdlib/v1.10/Pkg/src/Types.jl:70
 [2] find_project_file(env::Nothing)
   @ Pkg.Types /Applications/Julia-1.10.app/Contents/Resources/julia/share/julia/stdlib/v1.10/Pkg/src/Types.jl:220
 [3] Pkg.Types.EnvCache(env::Nothing)
   @ Pkg.Types /Applications/Julia-1.10.app/Contents/Resources/julia/share/julia/stdlib/v1.10/Pkg/src/Types.jl:356
 [4] EnvCache
   @ /Applications/Julia-1.10.app/Contents/Resources/julia/share/julia/stdlib/v1.10/Pkg/src/Types.jl:356 [inlined]
 [5] Context
   @ /Applications/Julia-1.10.app/Contents/Resources/julia/share/julia/stdlib/v1.10/Pkg/src/Types.jl:406 [inlined]
 [6] status(pkgs::Vector{Pkg.Types.PackageSpec}; io::IOContext{Base.PipeEndpoint}, kwargs::@Kwargs{mode::Pkg.Types.PackageMode})
   @ Pkg.API /Applications/Julia-1.10.app/Contents/Resources/julia/share/julia/stdlib/v1.10/Pkg/src/API.jl:150
 [7] status(; name::Nothing, uuid::Nothing, version::Nothing, url::Nothing, rev::Nothing, path::Nothing, mode::Pkg.Types.PackageMode, subdir::Nothing, kwargs::@Kwargs{})
   @ Pkg.API /Applications/Julia-1.10.app/Contents/Resources/julia/share/julia/stdlib/v1.10/Pkg/src/API.jl:174
 [8] status()
   @ Pkg.API /Applications/Julia-1.10.app/Contents/Resources/julia/share/julia/stdlib/v1.10/Pkg/src/API.jl:165
 [9] top-level scope
   @ ~/Projects/quarto/quarto-playground/index.qmd:7

ERROR: Internal julia server error

Stack trace:
    at writeJuliaCommand (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/execute/julia.ts:663:13)
    at eventLoopTick (ext:core/01_core.js:153:7)
    at async executeJulia (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/execute/julia.ts:511:20)
    at async Object.execute (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/execute/julia.ts:137:16)
    at async renderExecute (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/command/render/render-files.ts:232:25)
    at async renderFileInternal (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/command/render/render-files.ts:592:35)
    at async renderFiles (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/command/render/render-files.ts:325:9)
    at async render (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/command/render/render-shared.ts:102:18)
    at async renderForPreview (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/command/preview/preview.ts:422:24)
    at async render (file:///Users/mcanouil/Projects/quarto/quarto-cli/src/command/preview/preview.ts:166:22)
```

From:

````qmd
---
format: html
engine: julia
---

```{julia}
Pkg.status()
```

```{julia}
#| eval: false
using Plots
plot(sin, 0, 2π)
```
````

**Reply by MichaelHatherly** [reply]
Thanks, I'll try reproduce this locally myself, it certainly shouldn't be doing

```
@ /Users/mcanouil/Projects/quarto/quarto-playground/index.qmd:6
no active project
Stacktrace:
 [1] pkgerror(msg::String)
```

Could you provide the full contents of that `Project.toml` and `Manifest.toml` so that I can be sure to have the exact same environment as you.

**Reply by mcanouil** [reply]
I think I'm into something.
There's might be conflicts of some sort.

**Reply by mcanouil** [reply]
Is there some kind of cache associated with Julia environments and/or the Julia engine?
The `index.qmd`, `Project.toml`, and `Manifest.toml` are exactly the same.
(the other files are not used in this context and are there only to setup this "project" to help debugging things from issues/discussions on Quarto)

The below screenshots are from Terminal, thus IDEs are not involved at all.

<img width="1512" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/37bcb737-6ebe-4875-bdfe-0cb704f175a7">

**Reply by MichaelHatherly** [reply]
> Is there some kind of cache associated with Julia environments and/or the Julia engine?

Not as such.

The `QuartoNotebookRunner` keeps the worker processes alive that run notebooks for 5mins of inactivity before killing them, so there is some state there. Workers are keyed by their file name, so changing the name of `/Users/mcanouil/Projects/quarto/quarto-playground/index.qmd` to something else would start a different process to run it, for example.

**Reply by jkrumbiegel** [reply]
I've only ever encountered `no active project` when I started julia with an empty load path, so it couldn't find anything to load. Maybe you could try running `LOAD_PATH` in the first cell?

**Reply by mcanouil** [reply]
Somehow the worker is still alive then, as if I rename to `index2.qmd`, then it works.
Note, that I played with the native engine from the "start".

**Reply by mcanouil** [reply]
DO you know how to kill that worker? (I sent my comment too early by mistake)

**Reply by MichaelHatherly** [reply]
> Somehow the worker is still alive then

@jkrumbiegel, perhaps the timeouts aren't quite working as expected?

**Reply by mcanouil** [reply]
> > Somehow the worker is still alive then
> 
> @jkrumbiegel, perhaps the timeouts aren't quite working as expected?

Maybe a remnant of the development version (I tried the PR with the Julia engine before it was merged)

> I've only ever encountered no active project when I started julia with an empty load path, so it couldn't find anything to load. Maybe you could try running LOAD_PATH in the first cell?

<img width="1624" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/16a394f9-105f-4714-adfe-02d0831075b1">

**Reply by jkrumbiegel** [reply]
> perhaps the timeouts aren't quite working as expected?

Could be, there are some tests on the QuartoNotebookRunner side but not integration tests with quarto

**Reply by jkrumbiegel** [reply]
> DO you know how to kill that worker?

sorry, missed that question:

Using `--execute-daemon-restart` with the quarto render command or setting `daemon: false` for one invocation under the `execute` key in the frontmatter will restart the worker.

You cannot kill the main server process unless you find out its pid with `--execute-debug` and kill it yourself

**Reply by mcanouil** [reply]
Restarting the daemon did the trick, thanks!

I'll keep that in mind as some users might need this when troubleshooting issues with the Julia engine.

**Reply by mcanouil** [reply, marked as answer]
I'll probably make a PR to improve the documentation and ping you there for a review.

For the time being:

- https://github.com/quarto-dev/quarto-cli/issues/9561

Thanks for the help!

**Comment by dmi3kno** [comment]
I am developing a Julia package and I want to use Plots in vignette (README.qmd), but my package does not need to depends on Plots. How do I make it explicit, that I am relying on system installation (`@v1.10`) of Plots, but the rest of the packages should be used from local package environment? Sort of like "Suggests" in R.

**Reply by cderv** [reply]
@dmi3kno I am not sure how this is relating to Quarto. Can you share more example ? You could open a new discussion as it seems quite specific  and not directly related to this initial discussion. 

We've recently updated our doc at  https://quarto.org/docs/computations/julia.html#notebook-environments where `--project` flag can be set to indicate QuartoNotebookRunner.jl to use a specific project while running the Julia cells in a qmd document. 

I would think setting that in your README.qmd would be enough to use a different environment. 

Let's discuss in a new thread maybe.

**Reply by dmi3kno** [reply]
It's probably not related to Quarto, I agree. It is a question about Julia environments, potentially connected to Quarto by `--project=...` exeflag. It feels like I want to execute notebook in an environment where both Plots and my new package are installed, but, unfortunately, the way I do it now (running README.qmd from the root of the package) means that Project.toml has to contain Plots. Perhaps the best practice for package development is to create a new environment in, say, `docs` subfolder and add both my newly developed package and Plots there. I need to think a little about it. I will open an issue when I am ready to ask a meaningful question.
